src/main/java/org/breedinginsight/services/ProgramUploadService.java

```diff
@@ -0,0 +1,233 @@
+        } catch(JsonProcessingException e) {
+            log.error(e.getMessage());
+            throw new UnprocessableEntityException("Problem converting traits json");
+        }
+
+        // delete any existing records for traits since we only want to allow one at a time
+        // if there is some failure in writing the new, the old will be wiped out but that's ok
+        // because by making the PUT call the client already expected an overwrite
+        programUploadDao.deleteUploads(programId, actingUser.getId(), UploadType.TRAIT);
+
+        // do not autopopulate fields, that will be done on actual trait creation
+        BatchUploadEntity uploadEntity = BatchUploadEntity.builder()
+                .type(UploadType.TRAIT)
+                .programId(programId)
+                .userId(actingUser.getId())
+                .data(JSONB.valueOf(json))
+                .createdBy(actingUser.getId())
+                .updatedBy(actingUser.getId())
+                .build();
+
+        // Insert and update
+        programUploadDao.insert(uploadEntity);
+        return programUploadDao.getUploadById(uploadEntity.getId()).get();
+    }
+
+    private void checkRequiredTraitFields(Trait trait) throws UnprocessableEntityException {
+
+        Method method = trait.getMethod();
+        Scale scale = trait.getScale();
+
+        if (method == null) {
+            throw new UnprocessableEntityException("Missing method");
+        }
+
+        if (scale == null) {
+            throw new UnprocessableEntityException("Missing scale");
+        }
+
+        if (isBlank(trait.getTraitName())) {
+            throw new UnprocessableEntityException("Missing trait name");
+        }
+        if (isBlank(trait.getDescription())) {
+            throw new UnprocessableEntityException("Missing trait description");
+        }
+        if (trait.getProgramObservationLevel() == null || isBlank(trait.getProgramObservationLevel().getName())) {
+            throw new UnprocessableEntityException("Missing trait level");
+        }
+        if (isBlank(method.getMethodName())) {
+            throw new UnprocessableEntityException("Missing method name");
+        }
+        if (isBlank(method.getDescription())) {
+            throw new UnprocessableEntityException("Missing method description");
+        }
+        if (isBlank(method.getMethodClass())) {
+            throw new UnprocessableEntityException("Missing method class");
+        }
+        if (isBlank(scale.getScaleName())) {
+            throw new UnprocessableEntityException("Missing scale name");
+        }
+        if (scale.getDataType() == null) {
+            throw new UnprocessableEntityException("Missing scale type");
+        }
+    }
+
+    private boolean isBlank(String field) {
+        if (field == null || field.isBlank()) {
+            return true;
+        }
+        return false;
+    }
+
+    private void checkTraitDataConsistency(Trait trait) throws UnprocessableEntityException {
```

Comment:
Ya, something like that is what I was envisioning as well once we got to the POST traits since it would be used there as well. I can split it out now though so that we have it.

Reply:
Sounds good

Reply:
Pushed changes.